// @progress/kendo-angular-treeview/dist/npm/hierarchy-binding.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var kendo_common_1 = require("@progress/kendo-common");
var data_bound_component_1 = require("./data-bound-component");
var utils_1 = require("./utils");
var rxjs_1 = require("rxjs");
var hierarchy_editing_service_1 = require("./drag-and-drop/editing-services/hierarchy-editing.service");
var utils_2 = require("./utils");
var treeview_filter_settings_1 = require("./treeview-filter-settings");
var drag_and_drop_directive_1 = require("./drag-and-drop/drag-and-drop.directive");
var default_callbacks_1 = require("./default-callbacks");
var mapToWrappers = function (currentLevelNodes, childrenField) {
    if (!utils_1.isArrayWithAtLeastOneItem(currentLevelNodes)) {
        return [];
    }
    return currentLevelNodes.map(function (node) { return ({
        dataItem: node,
        children: mapToWrappers(kendo_common_1.getter(childrenField)(node), childrenField)
    }); });
};
var ɵ0 = mapToWrappers;
exports.ɵ0 = ɵ0;
/**
 * A directive which encapsulates the retrieval of child nodes.
 */
var HierarchyBindingDirective = /** @class */ (function () {
    function HierarchyBindingDirective(treeView, dragAndDropDirective) {
        var _this = this;
        this.treeView = treeView;
        this.dragAndDropDirective = dragAndDropDirective;
        /**
         * Stores the unfiltered nodes
         */
        this.filterData = [];
        this.visibleNodes = new Set();
        this._filterSettings = treeview_filter_settings_1.defaultTreeviewFilterSettings;
        var shouldFilter = !utils_1.isPresent(this.dragAndDropDirective);
        this.treeView.isVisible = shouldFilter ? function (node) { return _this.visibleNodes.has(node); } : default_callbacks_1.isVisible;
    }
    Object.defineProperty(HierarchyBindingDirective.prototype, "childrenField", {
        /**
         * The field name which holds the data items of the child component.
         */
        get: function () {
            return this._childrenField;
        },
        /**
         * The field name which holds the data items of the child component.
         */
        set: function (value) {
            if (!value) {
                throw new Error("'childrenField' cannot be empty");
            }
            this._childrenField = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HierarchyBindingDirective.prototype, "nodes", {
        /**
         * The nodes which will be displayed by the TreeView.
         */
        set: function (values) {
            this.filterData = mapToWrappers(values, this.childrenField) || [];
            this.visibleNodes.clear();
            this.updateNodesVisibility(this.filterData);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HierarchyBindingDirective.prototype, "isVisible", {
        /**
         * @hidden
         * A callback which determines whether a TreeView node should be rendered as hidden.
         */
        set: function (fn) {
            this.treeView.isVisible = fn;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HierarchyBindingDirective.prototype, "filter", {
        /**
         * Applies a filter and changes the visibility of the component's nodes accordingly.
         * To customize the built-in filtering, use the [filterSettings]({% slug api_treeview_hierarchybindingdirective %}#toc-filtersettings) prop.
         */
        set: function (term) {
            this.handleFilterChange(term);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HierarchyBindingDirective.prototype, "filterSettings", {
        get: function () {
            return this._filterSettings;
        },
        /**
         * The settings which are applied when performing a filter on the component's data.
         */
        set: function (value) {
            this._filterSettings = tslib_1.__assign({}, treeview_filter_settings_1.defaultTreeviewFilterSettings, value);
        },
        enumerable: true,
        configurable: true
    });
    HierarchyBindingDirective.prototype.ngOnInit = function () {
        var _this = this;
        if (utils_1.isPresent(this.childrenField)) {
            this.treeView.children = function (item) { return rxjs_1.of(kendo_common_1.getter(_this.childrenField)(item)); };
            this.treeView.hasChildren = function (item) {
                var children = kendo_common_1.getter(_this.childrenField)(item);
                return Boolean(children && children.length);
            };
            this.treeView.editService = new hierarchy_editing_service_1.HierarchyEditingService(this);
            this.treeView.filterChange.subscribe(this.handleFilterChange.bind(this));
            if (this.treeView.filter) {
                this.handleFilterChange(this.treeView.filter);
            }
        }
    };
    /**
     * @hidden
     */
    HierarchyBindingDirective.prototype.handleFilterChange = function (term) {
        if (!this.filterData) {
            return;
        }
        var filteredNodes = term ? utils_2.filterTree(this.filterData, term, this.filterSettings, this.treeView.textField) : this.filterData;
        this.visibleNodes.clear();
        this.updateNodesVisibility(filteredNodes);
    };
    HierarchyBindingDirective.prototype.updateNodesVisibility = function (items) {
        var _this = this;
        items.forEach(function (wrapper) {
            _this.visibleNodes.add(wrapper.dataItem);
            if (wrapper.children) {
                _this.updateNodesVisibility(wrapper.children);
            }
        });
    };
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], HierarchyBindingDirective.prototype, "childrenField", null);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Array),
        tslib_1.__metadata("design:paramtypes", [Array])
    ], HierarchyBindingDirective.prototype, "nodes", null);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Function])
    ], HierarchyBindingDirective.prototype, "isVisible", null);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], HierarchyBindingDirective.prototype, "filter", null);
    tslib_1.__decorate([
        core_1.Input(),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], HierarchyBindingDirective.prototype, "filterSettings", null);
    HierarchyBindingDirective = tslib_1.__decorate([
        core_1.Directive({ selector: '[kendoTreeViewHierarchyBinding]' }),
        tslib_1.__param(1, core_1.Optional()), tslib_1.__param(1, core_1.Host()),
        tslib_1.__metadata("design:paramtypes", [data_bound_component_1.DataBoundComponent, drag_and_drop_directive_1.DragAndDropDirective])
    ], HierarchyBindingDirective);
    return HierarchyBindingDirective;
}());
exports.HierarchyBindingDirective = HierarchyBindingDirective;
